File: src/main/java/blog/typed/javadsl/Greeter2.java
package blog.typed.javadsl;

import akka.typed.Behavior;
import akka.typed.javadsl.Actor;

public abstract class Greeter2 {

  // no instances of this class, it's only a name space for messages
  // and static methods
  private Greeter2() {
  }

  interface Command {
  }

  public static class Greet implements Command {
  }

  public static class WhoToGreet implements Command {
    public final String who;

    public WhoToGreet(String who) {
      this.who = who;
    }
  }


  public static Behavior<Command> greeterBehavior() {
    return greeterBehavior("hello");
  }

  private static Behavior<Command> greeterBehavior(String currentGreeting) {
    return Actor.immutable(Greeter2.Command.class)
      .onMessage(Greeter2.WhoToGreet.class, (ctx, msg) -> greeterBehavior("hello, " + msg.who))
      .onMessage(Greeter2.Greet.class, (ctx, msg) -> {
        System.out.println(currentGreeting);
        return Actor.same();
      })
      .build();
  }

}

File: src/main/scala/blog/typed/scaladsl/Greeter2.scala
package blog.typed.scaladsl

import scala.io.StdIn

import akka.actor.typed.ActorSystem
import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.Actor
import akka.actor.typed.ActorRef

object Greeter2 {
  sealed trait Command
  case object Greet extends Command
  final case class WhoToGreet(who: String) extends Command

  val greeterBehavior: Behavior[Command] = greeterBehavior(currentGreeting = "hello")

  private def greeterBehavior(currentGreeting: String): Behavior[Command] =
    Actor.immutable[Command] { (ctx, msg) =>
      msg match {
        case WhoToGreet(who) =>
          greeterBehavior(s"hello, $who")
        case Greet =>
          println(currentGreeting)
          Actor.same
      }
    }
}

object HelloWorldApp2 {
  def main(args: Array[String]): Unit = {
    val root = Actor.deferred[Nothing] { ctx =>
      import Greeter2._
      val greeter: ActorRef[Command] = ctx.spawn(greeterBehavior, "greeter")
      greeter ! WhoToGreet("World")
      greeter ! Greet

      Actor.empty
    }
    val system = ActorSystem[Nothing](root, "HelloWorld")
    try {
      println("Press ENTER to exit the system")
      StdIn.readLine()
    } finally {
      system.terminate()
    }
  }
}
